README.md
# microfft

microfft is a library for computing fast fourier transforms that targets
embedded systems. It provides an in-place implementation of the Radix-2 FFT
algorithm. All computations are performed directly on the input buffer and
require no additional allocations. This makes microfft suitable for `no_std`
environments, like microcontrollers.

Speed is achieved mainly by maintaining a pre-computed sine table that is used
to look up the necessary twiddle factors. By replacing arithmetic operations
with simple memory lookups, we reduce the number of CPU cycles spent.
Unfortunately, the pre-computed table also claims a considerable amount of
memory, which might be a deal-breaker for some embedded projects (see
[Memory Usage](#memory-usage)).

microfft also implements a specialized algorithm for FFTs on real (instead
of complex) values. Naively one would calculate a real FFT simply by converting
the input to complex values (leaving the imaginary part empty) and running a
CFFT. microfft's RFFT algorithm instead packs pairs of real values into
a single complex one each, then computes a CFFT of half the original input
size, followed by some recombination magic. This has the effect of roughly
halving the number of CPU cycles required, as can be seen in the
[benchmark results][bench/README.md].

## Example

The following example demonstrates computing a 16-point RFFT on a set of
samples generated from a sine signal:

```rust
use std::convert::TryInto;
use std::f32::consts::PI;

// generate 16 samples of a sine wave at frequency 3
let sample_count = 16;
let signal_freq = 3.;
let sample_interval = 1. / sample_count as f32;
let mut samples: Vec<_> = (0..sample_count)
    .map(|i| (2. * PI * signal_freq * sample_interval * i as f32).sin())
    .collect();

// compute the RFFT of the samples
let mut samples: [_; 16] = samples.try_into().unwrap();
let spectrum = microfft::real::rfft_16(&mut samples);
// since the real-valued coefficient at the Nyquist frequency is packed into the
// imaginary part of the DC bin, it must be cleared before computing the amplitudes
spectrum[0].im = 0.0;

// the spectrum has a spike at index `signal_freq`
let amplitudes: Vec<_> = spectrum.iter().map(|c| c.norm() as u32).collect();
assert_eq!(&amplitudes, &[0, 0, 0, 8, 0, 0, 0, 0]);
```

## Requirements

Requires Rust version **1.56.0** or newer.

## Sine Tables

microfft keeps a single sine table to calculate the twiddle factors for all
FFT sizes. This removes some memory overhead compared to keeping a separate
table for each FFT size, as there would be duplication between those tables.

The default sine table supports FFTs up to size 4096. If you only want to
compute FFTs of smaller sizes, it is recommended to select the appropriate
`size-*` feature, to not waste memory. For example, if your maximum FFT size is
1024, add this to your `Cargo.toml`:

```toml
[dependencies.microfft]
default-features = false
features = ["size-1024"]
```

This tells microfft to not provide functions for computing FFTs of sizes larger
than 1024 and to keep only the 1024-point sine table.

If you want to compute FFTs with more than 4096 points, you also need to enable
the respective feature. In this case, disabling the default features is not
required as microfft always determines the sine table size based on the largest
size requested. So this works as expected:

```toml
[dependencies.microfft]
features = ["size-8192"]
```

## Bit-reversal Tables

The optional feature `bitrev-tables` enables the use of pre-computed tables of
bit-reversed indices required for the reordering of input values performed at
the start of each FFT. If this feature is disabled (the default), the
bit-reversals are computed at runtime instead.

Note that enabling bitrev tables significantly increases the memory usage of
microfft. While it can speed up FFT computation on some systems, there are also
architectures that provide dedicated bit-reversal instructions (like `RBIT` on
ARMv7). On such architectures, switching on bitrev tables is usually
detrimental to performance.

## `std` Usage

microfft provides a `std` feature meant to make the library more useful for
applications that can make use of the Rust standard library. Currently the only
thing it does is transitively enabling the `std` feature of the `num-complex`
crate, thereby making more methods available on the `Complex32` values returned
by the FFT functions.

As embedded applications usually run on targets that don't have a Rust standard
library, the `std` feature is disabled by default. You can enable it in your
`Cargo.toml`:

```toml
[dependencies.microfft]
features = ["std"]
```

## Limitations

microfft has a few limitations, mostly due to its focus on speed, that might
make it unsuitable for some embedded projects. You should know about these
if you consider using this library:

### Memory Usage

The use of pre-computed sine and bitrev tables means that microfft has
considerable requirements on read-only memory. If your chip doesn't have much
flash to begin with, this can be an issue.

The amount of memory required for tables depends on the the configuration of
the [`size-*`](#sine-tables) and [`bitrev-tables`](#bit-reversal-tables)
features:

| `size-*`     | without `bitrev-tables` | with `bitrev-tables` |
| ------------ | ----------------------: | -------------------: |
| `size-4`     |                       0 |                    8 |
| `size-8`     |                       4 |                   20 |
| `size-16`    |                      12 |                   44 |
| `size-32`    |                      28 |                   92 |
| `size-64`    |                      60 |                  188 |
| `size-128`   |                     124 |                  380 |
| `size-256`   |                     252 |                  764 |
| `size-512`   |                     508 |                1,532 |
| `size-1024`  |                   1,020 |                3,068 |
| `size-2048`  |                   2,044 |                6,140 |
| `size-4096`  |                   4,092 |               12,284 |
| `size-8192`  |                   8,188 |               24,572 |
| `size-16384` |                  16,380 |               49,148 |

In addition, the code size also increases with FFT size.

### Supported FFT Sizes

microfft only supports FFT point-sizes that are powers of two, a limitation of
the Radix-2 algorithm. Additionally, the maximum supported size is currently
16384, although this limit can be increased in the future as necessary.

### `f64` Support

This library currently only supports single-precision floating-point inputs.
Similarly to the FFT size limit, this is a restriction that might be lifted
in the future, should the need arise.

## License

This project is licensed under the MIT license ([LICENSE](LICENSE) or
http://opensource.org/licenses/MIT).

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in microfft by you, shall be licensed as above, without any
additional terms or conditions.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## 0.5.0 (2022-06-19)

### Changed

- **Breaking:** The MSRV has been increased to 1.56.0.
- **Breaking:** The former `maxn-*` features of this crate have been renamed
  to `size-*`. They still work basically the same way but are now additive,
  which means it is allowed that more than one of them is enabled at the same
  time. This makes microfft work as expected in the face of Cargo's feature
  unification.

### Added

- A `std` feature that is disabled by default. It transitively enables the
  `std` feature on the `num-complex` dependency, enabling more methods on the
  `Complex32` values returned by the FFT functions.
- Support for FFT sizes 8192 and 16384.


## 0.4.0 (2021-04-03)

### Changed

- **Breaking:** The MSRV has been increased to 1.51.0.
- **Breaking:** All API functions for computing FFTs now take references to
  arrays instead of references to slices. This has the benefit of moving the
  length checking to compile time, which makes those functions panic-proof.
- **Breaking:** When computing RFFTs, the real-valued coefficient at the
  Nyquist frequency is now packed into the imaginary part of the DC bin
  (bin 0). Previously this value was simply dropped from the RFFT output.
- Added `#[must_use]` annotations to all FFT API functions.

### Fixed

- Thanks to Cargo's new resolver it is now possible to build microfft as a
  stand-alone library.


## 0.3.1 (2020-11-09)

### Fixed

- Fixed a bug during the RFFT recombination calculation that caused a wrong
  output value in bin `N / 4`.


## 0.3.0 (2020-03-08)

### Changed

- Store only the largest sine table, instead of one for each FFT size. By
  default this is the table for the 4096-point FFT (the largest supported one),
  but smaller ones can be selected via the new `maxn-*` crate features.


## 0.2.0 (2020-03-08)

### Changed

- Bitrev tables are not used anymore by default, instead the bit-reversed
  indices are computed directly at runtime. This significantly reduces the
  memory usage of microfft. On architectures that provide a dedicated
  bit-reversal instruction (like `RBIT` on ARMv7), speed is also increased.
  The `bitrev-tables` feature can be enabled to still opt into using bitrev
  tables.


## 0.1.2 (2020-03-07)

### Changed

- Store pre-computed sine values instead of full twiddles, reducing the size
  of the twiddle tables to one fourth the prior size.

## 0.1.1 (2020-03-05)

### Added

- Support for FFT sizes 2048 and 4096.


## 0.1.0 (2020-03-04)

### Added

- Support for complex and real FFTs up to size 1024.


src/complex.rs
//! FFT on complex inputs (CFFT)

use crate::{cfft::*, Complex32};

/// Perform an in-place 2-point CFFT.
///
/// # Example
///
/// ```
/// use microfft::{Complex32, complex::cfft_2};
///
/// let mut input = [Complex32::default(); 2];
/// let result = cfft_2(&mut input);
/// ```
#[inline]
#[must_use]
pub fn cfft_2(input: &mut [Complex32; 2]) -> &mut [Complex32; 2] {
    CFftN2::transform(input);
    input
}

/// Perform an in-place 4-point CFFT.
///
/// # Example
///
/// ```
/// use microfft::{Complex32, complex::cfft_4};
///
/// let mut input = [Complex32::default(); 4];
/// let result = cfft_4(&mut input);
/// ```
#[cfg(feature = "size-4")]
#[inline]
#[must_use]
pub fn cfft_4(input: &mut [Complex32; 4]) -> &mut [Complex32; 4] {
    CFftN4::transform(input);
    input
}

/// Perform an in-place 8-point CFFT.
///
/// # Example
///
/// ```
/// use microfft::{Complex32, complex::cfft_8};
///
/// let mut input = [Complex32::default(); 8];
/// let result = cfft_8(&mut input);
/// ```
#[cfg(feature = "size-8")]
#[inline]
#[must_use]
pub fn cfft_8(input: &mut [Complex32; 8]) -> &mut [Complex32; 8] {
    CFftN8::transform(input);
    input
}

/// Perform an in-place 16-point CFFT.
///
/// # Example
///
/// ```
/// use microfft::{Complex32, complex::cfft_16};
///
/// let mut input = [Complex32::default(); 16];
/// let result = cfft_16(&mut input);
/// ```
#[cfg(feature = "size-16")]
#[inline]
#[must_use]
pub fn cfft_16(input: &mut [Complex32; 16]) -> &mut [Complex32; 16] {
    CFftN16::transform(input);
    input
}

/// Perform an in-place 32-point CFFT.
///
/// # Example
///
/// ```
/// use microfft::{Complex32, complex::cfft_32};
///
/// let mut input = [Complex32::default(); 32];
/// let result = cfft_32(&mut input);
/// ```
#[cfg(feature = "size-32")]
#[inline]
#[must_use]
pub fn cfft_32(input: &mut [Complex32; 32]) -> &mut [Complex32; 32] {
    CFftN32::transform(input);
    input
}

/// Perform an in-place 64-point CFFT.
///
/// # Example
///
/// ```
/// use microfft::{Complex32, complex::cfft_64};
///
/// let mut input = [Complex32::default(); 64];
/// let result = cfft_64(&mut input);
/// ```
#[cfg(feature = "size-64")]
#[inline]
#[must_use]
pub fn cfft_64(input: &mut [Complex32; 64]) -> &mut [Complex32; 64] {
    CFftN64::transform(input);
    input
}

/// Perform an in-place 128-point CFFT.
///
/// # Example
///
/// ```
/// use microfft::{Complex32, complex::cfft_128};
///
/// let mut input = [Complex32::default(); 128];
/// let result = cfft_128(&mut input);
/// ```
#[cfg(feature = "size-128")]
#[inline]
#[must_use]
pub fn cfft_128(input: &mut [Complex32; 128]) -> &mut [Complex32; 128] {
    CFftN128::transform(input);
    input
}

/// Perform an in-place 256-point CFFT.
///
/// # Example
///
/// ```
/// use microfft::{Complex32, complex::cfft_256};
///
/// let mut input = [Complex32::default(); 256];
/// let result = cfft_256(&mut input);
/// ```
#[cfg(feature = "size-256")]
#[inline]
#[must_use]
pub fn cfft_256(input: &mut [Complex32; 256]) -> &mut [Complex32; 256] {
    CFftN256::transform(input);
    input
}

/// Perform an in-place 512-point CFFT.
///
/// # Example
///
/// ```
/// use microfft::{Complex32, complex::cfft_512};
///
/// let mut input = [Complex32::default(); 512];
/// let result = cfft_512(&mut input);
/// ```
#[cfg(feature = "size-512")]
#[inline]
#[must_use]
pub fn cfft_512(input: &mut [Complex32; 512]) -> &mut [Complex32; 512] {
    CFftN512::transform(input);
    input
}

/// Perform an in-place 1024-point CFFT.
///
/// # Example
///
/// ```
/// use microfft::{Complex32, complex::cfft_1024};
///
/// let mut input = [Complex32::default(); 1024];
/// let result = cfft_1024(&mut input);
/// ```
#[cfg(feature = "size-1024")]
#[inline]
#[must_use]
pub fn cfft_1024(input: &mut [Complex32; 1024]) -> &mut [Complex32; 1024] {
    CFftN1024::transform(input);
    input
}

/// Perform an in-place 2048-point CFFT.
///
/// # Example
///
/// ```
/// use microfft::{Complex32, complex::cfft_2048};
///
/// let mut input = [Complex32::default(); 2048];
/// let result = cfft_2048(&mut input);
/// ```
#[cfg(feature = "size-2048")]
#[inline]
#[must_use]
pub fn cfft_2048(input: &mut [Complex32; 2048]) -> &mut [Complex32; 2048] {
    CFftN2048::transform(input);
    input
}

/// Perform an in-place 4096-point CFFT.
///
/// # Example
///
/// ```
/// use microfft::{Complex32, complex::cfft_4096};
///
/// let mut input = [Complex32::default(); 4096];
/// let result = cfft_4096(&mut input);
/// ```
#[cfg(feature = "size-4096")]
#[inline]
#[must_use]
pub fn cfft_4096(input: &mut [Complex32; 4096]) -> &mut [Complex32; 4096] {
    CFftN4096::transform(input);
    input
}

/// Perform an in-place 8192-point CFFT.
///
/// # Example
///
/// ```
/// use microfft::{Complex32, complex::cfft_8192};
///
/// let mut input = [Complex32::default(); 8192];
/// let result = cfft_8192(&mut input);
/// ```
#[cfg(feature = "size-8192")]
#[inline]
#[must_use]
pub fn cfft_8192(input: &mut [Complex32; 8192]) -> &mut [Complex32; 8192] {
    CFftN8192::transform(input);
    input
}

/// Perform an in-place 16384-point CFFT.
///
/// # Example
///
/// ```
/// use microfft::{Complex32, complex::cfft_16384};
///
/// let mut input = [Complex32::default(); 16384];
/// let result = cfft_16384(&mut input);
/// ```
#[cfg(feature = "size-16384")]
#[inline]
#[must_use]
pub fn cfft_16384(input: &mut [Complex32; 16384]) -> &mut [Complex32; 16384] {
    CFftN16384::transform(input);
    input
}
